### ISSUE TITLE
[Bug]: GTA5.exe freezes after several script reloads

### ISSUE THREAD

**Issue by nomakewan** [scripthookvdotnet/scripthookvdotnet#1304]
### What modules are you seeing the problem on?

ScriptHookVDotNet.asi (runtime), ScriptHookVDotNet3.dll (v3 SDK/API)

### SHVDN Version

dde6861

### Game Version

v1.0.2944.0

### Bug Description

In all nightly SHVDN versions since commit dde6861, refreshing a script that performs a significant number of native calls will eventually cause GTA5.exe to crash, leaving an Event "The program GTA5.exe version 1.0.2944.0 stopped interacting with Windows and was closed." in the Event Viewer.

This is difficult to reproduce with tiny demo scripts that may only make a handful of calls. But for larger, more complex scripts I can cause the GTA5.exe crash to happen very reliably within a minute or so.

All versions prior to commit dde6861 do not exhibit this same behavior; I have refreshed them aggressively and have not been able to cause the program to crash.

I should note, additionally, that I do not have debug mode enabled in SHV; this is completely unrelated to #547 as it is only happening when I refresh SHVDN's scripts.

### Steps To Reproduce

1. Set ReloadKey to F5 in ScriptHookVDotNet.ini for ease.
2. Load into the game with a script that makes a lot of native calls.
3. Get into complex situations (enter Los Santos Customs with a script-controlled vehicle, drive down a busy highway, etc)
4. Refresh all scripts using F5.
5. Repeat 3-4; after a few refreshes, GTA5.exe will crash.

### Link To Code Example

https://github.com/bttfv/BackToTheFutureV

### The Current Behavior

GTA5.exe crashes, raising a prompt that the program must be terminated and leaving an event in the Event Viewer stating as much.

### The Expected Behavior

GTA5.exe cannot crash simply because SHVDN reloaded its scripts.

**Comment by kagikn** [comment]
I haven't reproduce if you mean the exe does an immediate crash, but I saw the game indefinitely freeze in 5d98746 when I reloaded SHVDN with BTTFV in Los Santos Customs. I suspect SHVDN can suffer a deadlock when reloading.
Did you omit the term "freeze" because it doesn't sound too important? If so, please distinguish indefinite freezes and immediate crashes when reporting, it may make issues harder to diagnose.

**Comment by kagikn** [comment]
Umm, seems that `AppDomain.Unload` may not return now infinitely. I didn't had issues around that method, looks like we need to gather a lot of info how it works and info of pitfalls.

**Comment by nomakewan** [comment]
> I haven't reproduce if you mean the exe does an immediate crash, but I saw the game infinitely freeze in [5d98746](https://github.com/scripthookvdotnet/scripthookvdotnet/commit/5d987460ff847a0d0dadc717b6353c793ce5a78b) when I reloaded SHVDN with BTTFV in Los Santos Customs. Couldn't repro the issue in [1f001b2](https://github.com/scripthookvdotnet/scripthookvdotnet/commit/1f001b264ad7cd03d8e378a674087221da4862cf) as you suggested. ~I suspect SHVDN can suffer a deadlock (by a lock for unloading stuff) when reloading. If that is really the problem, maybe we just couldn't reproduce it when #1192 was merged.~ Doesn't seem that unload lock is the problem. Needs more investigation. Did you omit the term "freeze" because it doesn't sound too important? If so, please distinguish infinite freezes and immediate crashes when reporting, it may make issues harder to diagnose. I'm asking since you reported a event with the string "stopped interacting with Windows and was closed.", which exception crashes won't report, such as access violation exceptions.

I have created a video showing the issue.

[Video](https://www.youtube.com/watch?v=YiQexS_1Pbs)

In this video it actually only took two refreshes of SHVDN to cause the crash to happen. Normally I have to work a little harder. Also the delay between 0:40 and 1:29 in the video is because Window Capture in OBS was not showing the crash condition, so I had to switch over to OBS to get a different capture input up so that it would be visible. After 0:40, feel free to skip to 1:29 to see the crash condition.

**Comment by kagikn** [comment]
Sorry for late reply than usual. Yes, my guess is right in if this issue is about freezing the app/exe, not crashing.

Are you sure BTTFV or any its dependencies won't call some synchronization APIs in their finalizer methods so they won't make the finalizer thread get deadlocked? I'm struggling if the issue you reported is really because SHVDN has some flaws, but I'm wondering one thing. Which is, SHVDN always wait for `NtWaitForMultipleObjects` or some `WaitForMultipleObjects` thingy, which could mean some synchronization API is called under the hood in a finalizer. I also find [an article about deadlock in finalizers in .NET](https://dzone.com/articles/sta-objects-and-finalizer).

And SHVDN seems to get blocked before the `AppDomain.Unload` call in `ScriptDomain.Unload` at the 2nd reload when BTTFV is loaded and this snippet is inserted before `AppDomain.Unload` call in `ScriptDomain.Unload`... So all finalizer methods on unreachable objects should be called for testing.
```cs
GC.Collect();
GC.WaitForPendingFinalizers();
GC.Collect();
GC.WaitForPendingFinalizers();
```

In short, kind of sorry to say, but I'm doubting if this issue is SHVDN's fault. All of finalizer methods of SHVDN basically just calls `Marshal.FreeCoTaskMem`, which shouldn't call `WaitForMultipleObjects`.

**Comment by nomakewan** [comment]
Thank you for the response!

As I am not certain that any of the dependencies do this, I will start testing to see if I can tease out what is going on. Our dependencies are fairly small; there's BTTFV itself, FusionLibrary (our helper library), KlangRageAudioLibrary (the helper library to interface with irrKlang.NET4) and irrKlang.NET4. The only other dependency is LemonUI but I'm assuming this issue wouldn't come from that.

Thank you very much for your response, and I will let you know how the testing goes as soon as I have more information to provide.

**Comment by nomakewan** [comment]
I will be closing this issue as it is indeed an issue with one of the dependencies. As a test I removed all of our sound engine references and the freeze vanished. After adding it back in the freeze returned. By then going into KlangRageAudioLibrary and commenting out the Disposal calls in AudioEngine.cs and AudioPlayer.cs, the freeze vanished.

So this appears to be some sort of race condition related to the disposal of objects by KlangRageAudioLibrary and not in fact a problem with SHVDN, as you suspected. It just didn't happen before because the rules on threads and such weren't being enforced properly.

Thank you again!
